### PR TITLE
Add a new script to check model testers' config

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Add a new check to check model testers will give tiny config.

The objective is to ensure no test will run with large configuration values (such as `BridgeTowerModelTest(er)`, except the integration or slow tests.

The check is not added to PR/daily CI workflow: we don't want to add more burden to contributors.

### The effect
Screenshot 2023-03-09 212533

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -158,6 +158,7 @@ jobs:
                       - v0.5-repository_consistency
             - run: pip install --upgrade pip
             - run: pip install .[all,quality]
+            - run: pip install pytest timeout_decorator parameterized
```
Comment (ydshieh) [inline .circleci/config.yml:161]: These extra 3 dependencies will allow all PyTorch model tester classes to be checked.

File: utils/check_model_tester.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    model_types = [x for x in model_types if x not in ["auto", "__pycache__"]]
+
+    for model_type in model_types[:]:
+        test_dir = os.path.join(model_test_root_dir, model_type)
+        test_fns = os.listdir(test_dir)
+        test_fns = [x for x in test_fns if x.startswith("test_modeling_")]
+        # TODO: deal with TF/Flax too
+        test_fns = [
+            x for x in test_fns if not (x.startswith("test_modeling_tf_") or x.startswith("test_modeling_flax_"))
+        ]
```
Comment (ydshieh) [inline utils/check_model_tester.py:38]: So far just check PyTorch test files. Can extend to TF/Flax later if we want.

File: utils/check_model_tester.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                try:
+                    tester = tester_class(parent=None)
+                except Exception:
+                    continue
+                if hasattr(tester, "get_config"):
+                    config = tester.get_config()
+                    for k, v in config.to_dict().items():
+                        if isinstance(v, int) and v >= 100:
+                            # Need to deal with (allow) special cases, otherwise change some values
+                            failures.append(
```
Comment (ydshieh) [inline utils/check_model_tester.py:55]: Currently quite (a lot) of failures.

**Comment by ydshieh** [comment]
Request a review from @amyeroberts to celebrate her new core maintainer role 🔥 🚀 
:

File: utils/check_model_tester.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        test_dir = os.path.join(model_test_root_dir, model_type)
+        test_fns = os.listdir(test_dir)
+        test_fns = [x for x in test_fns if x.startswith("test_modeling_")]
+        # TODO: deal with TF/Flax too
+        test_fns = [
+            x for x in test_fns if not (x.startswith("test_modeling_tf_") or x.startswith("test_modeling_flax_"))
+        ]
+        for test_fn in test_fns:
+            test_file = os.path.join(test_dir, test_fn)
+            tester_classes = get_tester_classes(test_file)
```
Comment (ydshieh) [inline utils/check_model_tester.py:41]: Good example usage of new methods introduced in the PR #21856

**Review by sgugger — COMMENTED** [review]
I think that's way too specific and will require lots of exceptions. It would be more helpful to have something in a PR added a new model that would extract the time the tests took (as reported in the artifacts) and put it somewhere clearly visible, for instance the comment regarding the documentation.

**Comment by ydshieh** [comment]
> I think that's way too specific

I agree that `100` is way to specific.

> and will require lots of exceptions

Not really a lot of exceptions. So far, there are `245` errors given by this new check.
- `max_position_embeddings`: 78 places --> this is better to be changed to smaller values
- `hidden_size` or `xxx_dim`: 41 places --> also better to be changed
- `xxx_size` (other than `hidden_size`): 39 places  --> I guess better to change
- `xxx_token_id`: 19 places --> this could be skipped (they look weird, but it doesn't really matter)
- **we can skip other places for now**

> It would be more helpful to have something in a PR added a new model that would extract the time the tests took (as reported in the artifacts) and put it somewhere clearly visible, for instance the comment regarding the documentation.

The `test_torch` job of PR #20775 (Add BridgeTower model) took `37` minutes, while it took about `33` minutes on nightly run one day before. So yes, this might be a good way to look. However:
- even if we show the timing, **without the timing of the previous (full) run**, no one really knows if we should look into the timing/speed issue, and we/contributor just would not pay attention
  - try to grab  (automatically) **without the timing of the previous (full) run** seems to me not a super easy task
- we will have to identity PRs that are new model addition PRs (use diff tool?) + we will have to show in PR comments
  - This, plus the necessity to grab the previous running time, seems to me requiring much more work than just simply fix things I mentioned in the first part above

So let me continue a bit and see what would happen!

File: utils/check_model_tester.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                # A few tester classes don't have `parent` parameter in `__init__`
+                # TODO: deal this better
+                try:
+                    tester = tester_class(parent=None)
+                except Exception:
+                    continue
+                if hasattr(tester, "get_config"):
+                    config = tester.get_config()
+                    for k, v in config.to_dict().items():
+                        if isinstance(v, int) and v >= 100:
```
Comment (amyeroberts) [inline utils/check_model_tester.py:53]: We might want another check here to iterate over lists, as some of these also determine how large a model can be e.g. `hidden_sizes`

**Review by amyeroberts — COMMENTED** [review]
Nice :) Having a check to keep the health of our tests is definitely a good idea! 

I agree with @sgugger that this might be overly prescriptive - particularly as some of the parameters as co-dependent: even if two values are low, another because of the architecture might have to be the multiple of these and be >100. 

For the iteration over the test files - it might be easier to use something like `glob.glob('tests/models/**/test_modeling_*.py')` and then filter from there.

File: utils/check_model_tester.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+if __name__ == "__main__":
+    failures = []
+
+    model_test_root_dir = os.path.join("tests", "models")
+
+    model_types = os.listdir(model_test_root_dir)
+    model_types = [x for x in model_types if os.path.isdir(os.path.join(model_test_root_dir, x))]
+    model_types = [x for x in model_types if x not in ["auto", "__pycache__"]]
+
+    for model_type in model_types[:]:
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply utils/check_model_tester.py:31]: yeah, the PR is incomplete yet, and I was using this slice to run with fewer cases

File: utils/check_model_tester.py (hunk 5 above)
Comment (ydshieh) [reply utils/check_model_tester.py:53]: nice catch!

**Comment by LysandreJik** [comment]
I also don't think this check is super valuable as it will add another burden on the contributors for something that really only affects us (we're the ones to handle CI, memory, timeouts, etc.).

I think aiming for tiny tiny models is great, but it's not the end of the world if a few slip through the cracks which we have to correct after.

**Comment by ydshieh** [comment]
> it will add another burden on the contributors

Yeah, that's a super valid point. I don't feel strong, so more than happy to close the PR.

But still give you a rough numbers: 

> if a few slip through the cracks which we have to correct after.

currently there are `245` places this check identified, a few examples

```bash
config["max_speech_positions"] = 4000 which is too large for testing!
config["num_block_records"] = 13353718 which is too large for testing!
config["max_2d_position_embeddings"] = 1024 which is too large for testing!
```

Without the check, more such cases will accumulate (but most of them are not as extreme as `BridgeModelTest` . Also maybe we will only have to deal them after a long long long period of time.

**Comment by ydshieh** [comment]
Well, although we are not going to add a check on PR CI, I  think it might be handy if we add the script - just in case when we need to perform housekeeping. It's not complex this new script, but always better if it is already there when we need it.

I also change the check to just a few major attributes for now.

@sgugger @LysandreJik Let me know if you are happy with this addition (without it being run on PR).

File: utils/check_model_tester.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                        target = None
+                        if k in ["vocab_size"]:
+                            target = 100
+                        elif k in ["max_position_embeddings"]:
+                            target = 128
+                        elif k in ["hidden_size", "d_model"]:
+                            target = 40
+                        elif k == ["num_layers", "num_hidden_layers", "num_encoder_layers", "num_decoder_layers"]:
+                            target = 5
+                        if target is not None and v > target:
```
Comment (ydshieh) [inline utils/check_model_tester.py:56]: Just check a few most well know attributes (for now).

File: utils/check_model_tester.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    test_files = glob.glob(pattern)
+    # TODO: deal with TF/Flax too
+    test_files = [
+        x for x in test_files if not (x.startswith("test_modeling_tf_") or x.startswith("test_modeling_flax_"))
+    ]
+
+    for test_file in test_files:
+        tester_classes = get_tester_classes(test_file)
+        for tester_class in tester_classes:
+            print(tester_class.__name__)
```
Comment (sgugger) [inline utils/check_model_tester.py:36]: Should be removed, no?

**Review by sgugger — APPROVED** [review]
Ok for me to keep the script accessible but not use it in the CI!

**Pull request merged**